src/cpp/r/session/RConsoleActions.cpp

```diff
@@ -85,29 +88,51 @@ void ConsoleActions::add(int type, const std::string& data)
       else if (type == kConsoleActionInput)
       {
          std::vector<std::string> input;
-         boost::algorithm::split(input,
-                                 data,
-                                 boost::algorithm::is_any_of("\n"));
+         boost::algorithm::split(input, data, boost::algorithm::is_any_of("\n"));
          pendingInput_.insert(pendingInput_.end(), input.begin(), input.end());
       }
+      else if (data.empty())
+      {
+         return;
+      }
 
-
-      // automatically combine consecutive output actions (up to 512 bytes)
-      // we enforce a limit so that the limit defined for our circular buffer
-      // (see setCapacity above) implies a content size limit as well (if we
-      // didn't cap the size of combined output then the output actions could
-      // grow to arbitrary size)
-      if (type == kConsoleActionOutput &&
-          actionsType_.size() > 0      &&
-          actionsType_.back().getInt() == kConsoleActionOutput &&
-          actionsData_.back().getString().size() < 512)
```

Comment:
Note that the previous code only set this restriction on the already-existing data, not the newly-incoming data, so a single large piece of data could still cause us to overflow here.